src/charm.py

```diff
@@ -0,0 +1,353 @@
+
+        self.framework.observe(self.on.fiveg_nrf_relation_joined, self._configure_sdcore_smf)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._configure_sdcore_smf)
+
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="pfcp", port=PFCP_PORT, protocol="UDP"),
+                ServicePort(name="sbi", port=SMF_SBI_PORT),
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+            ],
+        )
+
+        self._metrics_endpoint = MetricsEndpointProvider(
+            self,
+            jobs=[
+                {
+                    "static_configs": [{"targets": [f"*:{PROMETHEUS_PORT}"]}],
+                }
+            ],
+        )
+
+    def _on_install(self, event: InstallEvent) -> None:
+        """Handles the install event.
+
+        Args:
+            event (InstallEvent): Juju event.
+        """
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+        self._write_ue_config_file()
+
+    def _configure_sdcore_smf(self, event: EventBase) -> None:
+        """Adds pebble layer and manages Juju unit status.
+
+        Args:
+            event: Juju event
+        """
+        if not self._default_database_relation_is_created:
+            self.unit.status = BlockedStatus(
+                "Waiting for `default-database` relation to be created"
+            )
+            return
+        if not self._smf_database_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for `smf-database` relation to be created")
+            return
+        if not self._nrf_relation_is_created:
+            self.unit.status = BlockedStatus("Waiting for NRF relation to be created")
+            return
+        if not self._container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            return
+        if not self._default_database_is_available:
+            self.unit.status = WaitingStatus(
+                "Waiting for `default-database` relation to be available"
+            )
+            return
+        if not self._smf_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for `smf-database` relation to be available")
+            return
+        if not self._nrf_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF relation to be available")
+            return
+        if not self._ue_config_file_is_written:
+            event.defer()
+            self.unit.status = WaitingStatus(
+                f"Waiting for `{UEROUTING_CONFIG_FILE}` config file to be pushed to workload container"  # noqa: W505, E501
+            )
+            return
+        if not self._config_file_is_written:
```

Comment:
We'll also probably want to validate that:
- If the file is already created: the content matches what we expect it to be. This will be useful whenever any of the file parameter change (ex. NRF url).

Reply:
[Here's](https://github.com/canonical/sdcore-upf-operator/pull/1/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR195) an example of this.

Reply:
Done.